# Getting Your Data From Yelp!

In order to make sure you are on track to completing the project, you will complete this workbook first. Below are steps that you need to take in order to make sure you have your data from yelp and are ready to analyze it. Your cohort lead will review this workbook with you the Wednesday before your project is due.    

## Part 1 - Understanding your data and question

You will be pulling data from the Yelp API to complete your analysis. The api, however, provides you with a lot of information that will not be pertinent to your analysis. YOu will pull data from the api and parse through it to keep only the data that you will need. In order to help you identify that information,look at the API documentation and understand what data the api will provide you. 

Identify which data fields you will want to keep for your analysis. 

https://www.yelp.com/developers/documentation/v3/get_started

___

## Part 2 - Create ETL pipeline for the business data from the API

Now that you know what data you need from the API, you want to write code that will execute a api call, parse those results and then insert the results into the DB.  

It is helpful to break this up into three different functions (*api call, parse results, and insert into DB*) and then you can write a function/script that pull the other three functions together. 

Let's first do this for the Business endpoint.

- Write a function to make a call to the yelp API

In [1]:
import pandas as pd
import requests
import json

In [2]:
with open ('C:/Users/acumu/Documents/Flatiron/ds-east-042621-lectures/Phase_1/ds-apis-nbz32-main/.secrets/creds.json') as f:
    creds = json.load(f)

In [3]:
api_key = creds['key']
term = 'bike shop'
location = 'Washington DC'
SEARCH_LIMIT = 50
url_params = {
        'term': term.replace(' ', '+'),
        'location': location.replace(' ', '+'),
        'limit': SEARCH_LIMIT,
        'offset': 0 
    }

In [4]:
def yelp_call(url_params, api_key):
    # your code to make the yelp call
    url = 'https://api.yelp.com/v3/businesses/search'
    headers = {'Authorization': 'Bearer ' + api_key}
    
    data = requests.get(url, headers = headers, params = url_params)
    return data

In [5]:
results = yelp_call(url_params, api_key)

In [6]:
type(results)

requests.models.Response

In [7]:
results.json()

{'businesses': [{'id': 'astpI6075yghzNk2XFiybA',
   'alias': 'rollin-cycles-bicycle-shop-washington-2',
   'name': 'Rollin Cycles Bicycle Shop',
   'image_url': 'https://s3-media4.fl.yelpcdn.com/bphoto/ZCdE5Y9P82yXdF5LnPDHNw/o.jpg',
   'is_closed': False,
   'url': 'https://www.yelp.com/biz/rollin-cycles-bicycle-shop-washington-2?adjust_creative=dpB8uUsFa0hX6mpY12y4ew&utm_campaign=yelp_api_v3&utm_medium=api_v3_business_search&utm_source=dpB8uUsFa0hX6mpY12y4ew',
   'review_count': 103,
   'categories': [{'alias': 'bikes', 'title': 'Bikes'},
    {'alias': 'bike_repair_maintenance', 'title': 'Bike Repair/Maintenance'}],
   'rating': 3.5,
   'coordinates': {'latitude': 38.90779, 'longitude': -77.03236},
   'transactions': [],
   'price': '$$',
   'location': {'address1': '1314 14th St NW',
    'address2': 'Ste A',
    'address3': '',
    'city': 'Washington, DC',
    'zip_code': '20005',
    'country': 'US',
    'state': 'DC',
    'display_address': ['1314 14th St NW', 'Ste A', 'Washington

In [8]:
data_df = pd.DataFrame(results.json()['businesses'])

In [9]:
data_df

id                                              alias  \
0   astpI6075yghzNk2XFiybA            rollin-cycles-bicycle-shop-washington-2   
1   Dn04j5YabUDWS7nUbpfZGg                       the-daily-rider-washington-2   
2   F0fMNrrBU3OKMdwfCL5mRA                        upshift-workshop-washington   
3   YvDCWE1O_Q6JLMiKHb4tGQ                 city-bikes-adams-morgan-washington   
4   ATpgglOwwQUno5QhAEXc2g                           bicycle-space-washington   
5   OxnwyBMJN43a_5UHL3UHWQ                                 velofix-washington   
6   3tbIAR6xTrgSRO3m2da0Vg                      contes-bike-shop-washington-2   
7   2q9eO68Q-PDAUUfdx-zvqw                          the-bike-house-washington   
8   gmXiBdjkKVTGM-4mU8Sepw                      contes-bike-shop-washington-4   
9   yGJ5jSQPSqp8i5rojSJnHQ                      gearin-up-bicycles-washington   
10  NBsk50CXW85OFAkxnOao8Q                  district-cycle-works-washington-5   
11  obfGllLjyMkQ1iA_HtT1bQ                   erics-mobile-workshop-washington   
12  8eVGxeIopFn5hUX_YVkbGQ                    nomad-bicycleworks-washington-3   
13  R2HBjZ0HYWIRquHMVBNA_w  trek-bicycle-washington-d-c-georgetown-washing...   
14  8mwlRyZ6FH2VW10tlHE2Qg                        bicycle-pro-shop-washington   
15  2NvqroYky8n4wO6agru4Kw                   bikes-and-skates-capitol-heights   
16  Wqmx9LXrKzAUGeTZRV3XZg                   city-bikes-tenleytown-washington   
17  Nnfhnuqv2WTeHRs_mDFPRA                         big-wheel-bikes-alexandria   
18  jQoYuCkGdeGNRfif4JlYDw                        the-old-bike-shop-arlington   
19  dEUQ7Of8leYhaRF12qTswQ                         takoma-bicycle-takoma-park   
20  MnV2zaZSXToe-vpkTccZ4Q                      bike-and-roll-dc-washington-2   
21  -zIt1sAhZpndaiWjMFVGcA                           griffin-cycle-bethesda-2   
22  uq9j3G6MJtd_Mv5rfF5oRQ      pedego-electric-bikes-alexandria-alexandria-2   
23  Sb_3F7bcGYLGjUssQfiP6w                                   riide-washington   
24  CZSLn0w5xUESdprbfINn8w                      proteus-bicycles-college-park   
25  p9U34BqsfvgY2zKTW6fFkg                        electricitybikes-washington   
26  iILgX1Q8tc29Onj8OgdKOA                   bike-doctor-of-waldorf-waldorf-2   
27  cU4Dg9ltf53aFoLBh8Ndkw                        silver-cycles-silver-spring   
28  SVk-4xwL9tA9TSB0ZnWFBA                         big-wheel-bikes-washington   
29  RfsW8djTUV8TfDHy9TxfUw                                   rei-washington-4   
30  AjIOtN_0SlyY2dgVuBbwhg                     jafe-cycling-capitol-heights-2   
31  DPUPnCnPwCdcFl6U7nBNsw                       contes-bike-shop-arlington-2   
32  4wh6LsjL0Cki3IfJbSSzrA                 bicyclespace-washington-washington   
33  ky_3ss5asDpYqb8L1x3P-Q                   spokes-etc-bicycles-alexandria-6   
34  WXuilvVlx0MzBqX1e6GO6A                                rapha-dc-washington   
35  xSzKaAq6vgZZDn3S5-V9Hw                          arrow-bicycle-hyattsville   
36  q-pvnWBgielD--bA81hLFQ                         district-cycles-washington   
37  wdQepvuFBpPKJ1R6titiNA                          papillon-cycles-arlington   
38  NM9NmHZ_KbNsmrfrP3Whvg                   trek-bicycle-rockville-rockville   
39  1q-0mrzsr1EIF5U0aTPFXQ                            probike-fc-falls-church   
40  Z97inykgHENrsBTg1S__wA                      contes-bike-shop-washington-3   
41  YvguUf5tL6HDElwIAhjZ4Q                    wheel-nuts-bike-shop-alexandria   
42  CmJmSF3Cw0lxuxwusJV8_Q       trek-bicycle-arlington-clarendon-arlington-2   
43  xTrJ9Q5tNAAkj5p9Du7yRw            vélocity-bicycle-cooperative-alexandria   
44  AsOcjvsQQfADGDft5LmzpQ                              fleet-feet-washington   
45  3mFZjswCTJ2X4MRBNRdSjA                    the-bicycle-place-silver-spring   
46  htdy8C1Rdz9baITZemFqgg                        evo-cycle-works-rockville-4   
47  wuFuQt1lI0kvy8ZaCie8WQ                            phoenix-bikes-arlington   
48  0KQG1qvOX8w7lPWa61w8OQ                          big

In [ ]:
data_df.shape

In [ ]:
data_df.info()

In [ ]:
data_df.describe()

In [ ]:
pd.DataFrame(data_df.groupby('name'))

- Write a function to parse the API response so that you can easily insert the data in to the DB

In [ ]:
#data = results.json()

In [ ]:
#business = data['businesses'][0]

In [ ]:
#biz_info = (business['id'], business['name'], business['is_closed'], business['review_count'], business['rating'], business['location']['zip_code'])

In [ ]:
#biz_info

In [11]:
def parse_results(results):
    # your code to parse the result to make them easier to insert into the DB
    results = results.json()['businesses']
    parsed_data = []
    for business in results:
        if 'price' not in business:
            biz_info = (business['id'], business['name'], business['is_closed'],
                business['review_count'], business['rating'], 'NaN',
                business['location']['zip_code'])
            parsed_data.append(biz_info)
        else:
            biz_info = (business['id'], business['name'], business['is_closed'],
                business['review_count'], business['rating'], business['price'],
                business['location']['zip_code'])
            parsed_data.append(biz_info)
    return parsed_data

In [12]:
parsed_results = parse_results(results)
parsed_results

[('astpI6075yghzNk2XFiybA',
  'Rollin Cycles Bicycle Shop',
  False,
  103,
  3.5,
  '$$',
  '20005'),
 ('Dn04j5YabUDWS7nUbpfZGg', 'The Daily Rider', False, 85, 4.5, '$$', '20002'),
 ('F0fMNrrBU3OKMdwfCL5mRA', 'Upshift Workshop', False, 6, 5.0, 'NaN', '20001'),
 ('YvDCWE1O_Q6JLMiKHb4tGQ',
  'City Bikes - Adams Morgan',
  False,
  257,
  3.0,
  '$$$',
  '20009'),
 ('ATpgglOwwQUno5QhAEXc2g', 'Bicycle Space', False, 54, 4.0, '$$$', '20002'),
 ('OxnwyBMJN43a_5UHL3UHWQ', 'velofix', False, 39, 5.0, 'NaN', '20001'),
 ('3tbIAR6xTrgSRO3m2da0Vg',
  "Conte's Bike Shop",
  False,
  48,
  4.0,
  '$$',
  '20003'),
 ('2q9eO68Q-PDAUUfdx-zvqw', 'The Bike House', False, 7, 5.0, '$', '20542'),
 ('gmXiBdjkKVTGM-4mU8Sepw',
  "Conte's Bike Shop",
  False,
  7,
  3.5,
  'NaN',
  '20009'),
 ('yGJ5jSQPSqp8i5rojSJnHQ',
  "Gearin' Up Bicycles",
  False,
  24,
  4.5,
  '$',
  '20002'),
 ('NBsk50CXW85OFAkxnOao8Q',
  'District Cycle Works',
  False,
  12,
  4.5,
  'NaN',
  '20007'),
 ('obfGllLjyMkQ1iA_HtT1bQ',
  "E

In [13]:
bike_df = pd.DataFrame(parsed_results, columns = ('ID', 'Name', 'Is_Closed',
                                           'Review_Count', 'Rating', 'Price',
                                           'Zip_Code'))
bike_df


ID                                     Name  \
0   astpI6075yghzNk2XFiybA               Rollin Cycles Bicycle Shop   
1   Dn04j5YabUDWS7nUbpfZGg                          The Daily Rider   
2   F0fMNrrBU3OKMdwfCL5mRA                         Upshift Workshop   
3   YvDCWE1O_Q6JLMiKHb4tGQ                City Bikes - Adams Morgan   
4   ATpgglOwwQUno5QhAEXc2g                            Bicycle Space   
5   OxnwyBMJN43a_5UHL3UHWQ                                  velofix   
6   3tbIAR6xTrgSRO3m2da0Vg                        Conte's Bike Shop   
7   2q9eO68Q-PDAUUfdx-zvqw                           The Bike House   
8   gmXiBdjkKVTGM-4mU8Sepw                        Conte's Bike Shop   
9   yGJ5jSQPSqp8i5rojSJnHQ                      Gearin' Up Bicycles   
10  NBsk50CXW85OFAkxnOao8Q                     District Cycle Works   
11  obfGllLjyMkQ1iA_HtT1bQ                   Eric's Mobile Workshop   
12  8eVGxeIopFn5hUX_YVkbGQ                       Nomad Bicycleworks   
13  R2HBjZ0HYWIRquHMVBNA_w  Trek Bicycle Washington D.C. Georgetown   
14  8mwlRyZ6FH2VW10tlHE2Qg                         Bicycle Pro Shop   
15  2NvqroYky8n4wO6agru4Kw                           Bikes & Skates   
16  Wqmx9LXrKzAUGeTZRV3XZg                    City Bikes Tenleytown   
17  Nnfhnuqv2WTeHRs_mDFPRA                          Big Wheel Bikes   
18  jQoYuCkGdeGNRfif4JlYDw                        The Old Bike Shop   
19  dEUQ7Of8leYhaRF12qTswQ                           Takoma Bicycle   
20  MnV2zaZSXToe-vpkTccZ4Q                         Bike And Roll DC   
21  -zIt1sAhZpndaiWjMFVGcA                            Griffin Cycle   
22  uq9j3G6MJtd_Mv5rfF5oRQ         Pedego Electric Bikes Alexandria   
23  Sb_3F7bcGYLGjUssQfiP6w                                    Riide   
24  CZSLn0w5xUESdprbfINn8w                         Proteus Bicycles   
25  p9U34BqsfvgY2zKTW6fFkg                         ElectriCityBikes   
26  iILgX1Q8tc29Onj8OgdKOA                   Bike Doctor of Waldorf   
27  cU4Dg9ltf53aFoLBh8Ndkw                            Silver Cycles   
28  SVk-4xwL9tA9TSB0ZnWFBA                          Big Wheel Bikes   
29  RfsW8djTUV8TfDHy9TxfUw                                      REI   
30  AjIOtN_0SlyY2dgVuBbwhg                             Jafe Cycling   
31  DPUPnCnPwCdcFl6U7nBNsw                        Conte's Bike Shop   
32  4wh6LsjL0Cki3IfJbSSzrA                BicycleSPACE - Washington   
33  ky_3ss5asDpYqb8L1x3P-Q                     Spokes Etc. Bicycles   
34  WXuilvVlx0MzBqX1e6GO6A                                 Rapha DC   
35  xSzKaAq6vgZZDn3S5-V9Hw                            Arrow Bicycle   
36  q-pvnWBgielD--bA81hLFQ                          District Cycles   
37  wdQepvuFBpPKJ1R6titiNA                          Papillon Cycles   
38  NM9NmHZ_KbNsmrfrP3Whvg                   Trek Bicycle Rockville   
39  1q-0mrzsr1EIF5U0aTPFXQ                               ProBike FC   
40  Z97inykgHENrsBTg1S__wA                        Conte's Bike Shop   
41  YvguUf5tL6HDElwIAhjZ4Q                     Wheel Nuts Bike Shop   
42  CmJmSF3Cw0lxuxwusJV8_Q         Trek Bicycle Arlington Clarendon   
43  xTrJ9Q5tNAAkj5p9Du7yRw             VéloCity Bicycle Cooperative   
44  AsOcjvsQQfADGDft5LmzpQ                               Fleet Feet   
45  3mFZjswCTJ2X4MRBNRdSjA                        The Bicycle Place   
46  htdy8C1Rdz9baITZemFqgg                          Evo Cycle Works   
47  wuFuQt1lI0kvy8ZaCie8WQ                            Phoenix Bikes   
48  0KQG1qvOX8w7lPWa61w8OQ                          Big Wheel Bikes   
49  d50OtT7Y4N5l3OeBtRwIdA                          Big Wheel Bikes   

    Is_Closed  Review_Count  Rating Price Zip_Code  
0       False           103     3.5    $$    20005  
1       False            85     4.5    $$    20002  
2       False             6     5.0   NaN    20001  
3       False           257     3.0   $$$    20009  
4       False            54     4.0   $$$    20002  
5       False            39     5.0   NaN    20001  
6       False            48    

- Write a function to take your parsed data and add it to the csv file where you will store all of your results. 

In [18]:
bike_df.to_csv('C:/Users/acumu/Documents/Flatiron/Phase_1_Project/bikedata.csv', index = False, encoding='utf-8')


In [ ]:
api_key = creds['key']
term = 'bike shop'
location = 'Washington DC'
SEARCH_LIMIT = 50
url_params = {
        'term': term.replace(' ', '+'),
        'location': location.replace(' ', '+'),
        'limit': SEARCH_LIMIT,
        'offset': 50 
    }

In [ ]:
def yelp_call(url_params, api_key):
    # your code to make the yelp call
    url = 'https://api.yelp.com/v3/businesses/search'
    headers = {'Authorization': 'Bearer ' + api_key}
    
    data = requests.get(url, headers = headers, params = url_params)
    return data

In [ ]:
results2 = yelp_call(url_params, api_key)

In [ ]:
results2.json()

In [ ]:
data2_df = pd.DataFrame(results2.json()['businesses'])
data2_df

In [ ]:
data2_df.info()

In [ ]:
data3_df.info()

In [ ]:
def parse_results2(results):
    # your code to parse the result to make them easier to insert into the DB
    results = results2.json()['businesses']
    parsed_data = []
    for business in results:
        if 'price' not in business:
            biz_info = (business['id'], business['name'], business['is_closed'],
                business['review_count'], business['rating'], 'NaN',
                business['location']['zip_code'])
            parsed_data.append(biz_info)
        else:
            biz_info = (business['id'], business['name'], business['is_closed'],
                business['review_count'], business['rating'], business['price'],
                business['location']['zip_code'])
            parsed_data.append(biz_info)
    return parsed_data

In [ ]:
parsed_results = parse_results2(results)
parsed_results

In [ ]:
bike_df2 = pd.DataFrame(parsed_results, columns = ('ID', 'Name', 'Is_Closed',
                                           'Review_Count', 'Rating', 'Price',
                                           'Zip_Code'))
bike_df2


In [ ]:
bike_df2.to_csv('C:/Users/acumu/Documents/Flatiron/ds-east-042621-lectures/Phase_1/phase-1-project-yelp/bikedata2.csv', index = False)

In [ ]:
bike_df2.shape

In [ ]:
api_key = creds['key']
term = 'bike shop'
location = 'Washington DC'
SEARCH_LIMIT = 50
url_params = {
        'term': term.replace(' ', '+'),
        'location': location.replace(' ', '+'),
        'limit': SEARCH_LIMIT,
        'offset': 100 
    }

In [ ]:
def yelp_call(url_params, api_key):
    # your code to make the yelp call
    url = 'https://api.yelp.com/v3/businesses/search'
    headers = {'Authorization': 'Bearer ' + api_key}
    
    data = requests.get(url, headers = headers, params = url_params)
    return data

In [ ]:
results3 = yelp_call(url_params, api_key)

In [ ]:
results3.json()

In [ ]:
data3_df = pd.DataFrame(results2.json()['businesses'])
data3_df

In [ ]:
def parse_results3(results):
    # your code to parse the result to make them easier to insert into the DB
    results = results2.json()['businesses']
    parsed_data = []
    for business in results:
        if 'price' not in business:
            biz_info = (business['id'], business['name'], business['is_closed'],
                business['review_count'], business['rating'], 'NaN',
                business['location']['zip_code'])
            parsed_data.append(biz_info)
        else:
            biz_info = (business['id'], business['name'], business['is_closed'],
                business['review_count'], business['rating'], business['price'],
                business['location']['zip_code'])
            parsed_data.append(biz_info)
    return parsed_data

In [ ]:
parsed_results = parse_results3(results)
parsed_results

In [ ]:
bike_df3 = pd.DataFrame(parsed_results, columns = ('ID', 'Name', 'Is_Closed',
                                           'Review_Count', 'Rating', 'Price',
                                           'Zip_Code'))
bike_df3

In [ ]:
bike_df3.to_csv('C:/Users/acumu/Documents/Flatiron/ds-east-042621-lectures/Phase_1/phase-1-project-yelp/bikedata3.csv', index = False)

In [ ]:
bike_df3.shape

In [ ]:
 csv_filepath = 'C:/Users/acumu/Documents/Flatiron/Phase_1_Project/dcbike_df.csv'

In [ ]:
def df_safe(csv_filepath, parsed_results):
    # your code to open the csv file, concat the current data, and save the data. 
    parsed_results = pd.DataFrame(parsed_results)
    return parsed_results.to_csv(csv_filepath, mode='a', index = False, header = False)

- Write a script that combines the three functions above into a single process.

While it will take some experimentation to write the functions above, once you get them working it will be best to put them in a `.py` file and then import the functions to use in a script 

In [ ]:
from helper_funcs import *

In [ ]:


# create a variable  to keep track of which result you are in. 
cur = 0

#set up a while loop to go through and grab the result 
while cur < 200 and cur < 1000:
    #set the offset parameter to be where you currently are in the results (replace cur with number of results)
    url_params['offset'] = cur
    #make your API call with the new offset number
    results = yelp_call(url_params, api_key)
    
    #after you get your results you can now use your function to parse those results
    parsed_results = parse_results(results)
    
    # use your function to insert your parsed results into the db
    df_safe(csv_filepath, parsed_results)
    #increment the counter by 50 to move on to the next results
    cur += 50

___

## Part 3 -  Create ETL pipeline for the restaurant review data from the API

You've done this for the Businesses, now you need to do this for reviews. You will follow the same process, but your functions will be specific to reviews. Above you ahve a model of the functions you will need to write, and how to pull them together in one script. For this part, you ahve the process below 

- In order to pull the reveiws, you will need the business ids. So your first step will be to get all of the business ids from your businesses csv. 

- Write a function that takes a business id and makes a call to the API for reivews


- Write a function to parse out the relevant information from the reviews

- Write a function to save the parse data into a csv file containing all of the reviews. 

- Combine the functions above into a single script  

In [ ]:
business_ids = pd.read_csv('C:/Users/acumu/Documents/Flatiron/Phase_1_Project/dcbike_df.csv')
business_ids

In [ ]:
business_ids['ID']

In [ ]:
list(business_ids['ID'])

In [ ]:
api_key = creds['key']
term = 'bike shop'
location = 'Washington DC'
SEARCH_LIMIT = 50


In [ ]:
def yelp_call_review(api_key):
    reviews = []
    for ID in business_ids['ID']:
        url = 'https://api.yelp.com/v3/businesses/'+ID+'/reviews'
        headers = {'Authorization': 'Bearer ' + api_key}
        data = requests.get(url, headers = headers)
        reviews.append(data)
    return reviews


In [ ]:
url = 'https://api.yelp.com/v3/businesses/'+business_ids['ID'][0]+'/reviews'

In [ ]:
url

In [ ]:
results = yelp_call_review(api_key)
results

In [ ]:
type(results)

In [ ]:
def parse_results_review(results):
    # your code to parse the result to make them easier to insert into the DB
    results = results.json()['reviews']
    parsed_data = []
    for review in results:
        biz_info = (review['id'], review['text'], review['rating'], review['time_created'],
                   review['user']['name'])
        parsed_data.append(biz_info)
    return parsed_data


In [ ]:
parsed_results = parse_results_review(results)
parsed_results

In [ ]:
csv_review_filepath = 'C:/Users/acumu/Documents/Flatiron/Phase_1_Project/dcbike_review_df.csv'

In [ ]:
def df_safe_review(csv_review_filepath, parsed_results):
    # your code to open the csv file, concat the current data, and save the data. 
    parsed_results = pd.DataFrame(parsed_results)
    return parsed_results.to_csv(csv_review_filepath, mode='a', index = False, header = False)

In [ ]:
# create a variable  to keep track of which result you are in. 
cur = 0

#set up a while loop to go through and grab the result 
while cur < 200 and cur < 1000:
    #set the offset parameter to be where you currently are in the results (replace cur with number of results)
    url_params['offset'] = cur
    #make your API call with the new offset number
    results = yelp_call_review(url_params, api_key)
    
    #after you get your results you can now use your function to parse those results
    data = yelp_call_review(results)
    
    # use your function to insert your parsed results into the db
    df_safe_review(csv_review_filepath, parsed_results)
    #increment the counter by 50 to move on to the next results
    cur += 50

___

## Part 4 -  Using python and pandas, write code to answer the questions below. 


- Which are the 5 most reviewed businesses in your dataset?
- What is the highest rating recieved in your data set and how many businesses have that rating?
- What percentage of businesses have a rating greater than or  4.5?
- What percentage of businesses have a rating less than 3?
- What percentage of your businesseshave a price label of one dollar sign? Two dollar signs? Three dollar signs? No dollar signs?
- Return the text of the reviews for the most reviewed business. 
- Find the highest rated business and return text of the most recent review. If multiple business have the same rating, select the business with the most reviews. 
- Find the lowest rated business and return text of the most recent review.  If multiple business have the same rating, select the business with the least reviews. 


___

# Reference help

###  Pagination

Returning to the Yelp API, the [documentation](https://www.yelp.com/developers/documentation/v3/business_search) also provides us details regarding the API limits. These often include details about the number of requests a user is allowed to make within a specified time limit and the maximum number of results to be returned. In this case, we are told that any request has a maximum of 50 results per request and defaults to 20. Furthermore, any search will be limited to a total of 1000 results. To retrieve all 1000 of these results, we would have to page through the results piece by piece, retriving 50 at a time. Processes such as these are often refered to as pagination.

Now that you have an initial response, you can examine the contents of the json container. For example, you might start with ```response.json().keys()```. Here, you'll see a key for `'total'`, which tells you the full number of matching results given your query parameters. Write a loop (or ideally a function) which then makes successive API calls using the offset parameter to retrieve all of the results (or 5000 for a particularly large result set) for the original query. As you do this, be mindful of how you store the data. 

**Note: be mindful of the API rate limits. You can only make 5000 requests per day, and APIs can make requests too fast. Start prototyping small before running a loop that could be faulty. You can also use time.sleep(n) to add delays. For more details see https://www.yelp.com/developers/documentation/v3/rate_limiting.**

***Below is sample code that you can use to help you deal with the pagination parameter and bring all of the functions together.***


***Also, something might cause your code to break while it is running. You don't want to constantly repull the same data when this happens, so you should insert the data into the database as you call and parse it, not after you have all of the data***
